### PR TITLE
Implement h-feed discovery algorithm

### PR DESCRIPTION
**Pull request by capjamesg**
The microformats h-feed specification describes an algorithm to follow when finding the h-feed that represents a page.

This PR implements all of that algorithm, excluding the "implied h-feed" phase for which more clarification is needed before continuing.

### CODE REVIEW

File: src/indieweb_utils/feeds/discovery.py
Comment (jamesvandyne) [inline src/indieweb_utils/feeds/discovery.py]: If there is no `"type"` and the default parameter `[]` is returned, this code will throw an `IndexError` because you can't access item 0 of an empty array. Instead we should check that item.get("type") exists (and None will evaluate as False)  and then if the type of the first element equals "h-feed".

```suggestion
        h_feed = [item for item in parsed_feed["items"] if item.get("type") and item.get("type")[0] == "h-feed"]
```

File: src/indieweb_utils/feeds/discovery.py
Comment (jamesvandyne) [inline src/indieweb_utils/feeds/discovery.py]: I think we could use this function independently of discover_web_page_feeds. If we can run canonicalization on an already canonicalized url and get the same thing back, it's better to just try and re-canonicalize it so we're not making any assumptions about how this function will be used.

Comment (capjamesg) [reply src/indieweb_utils/feeds/discovery.py]: Can you elaborate on this suggestion? The discover_web_page_feeds function retrieves all of the feeds on a page, including the mf2 feed that someone is likely not to have discovered yet.

With regard to re-canonicalization, good call! I shall ad dspuport for an additional call to canonicalize_url in my next PR.

**Comment by capjamesg** [comment]
@jamesvandyne Can you do another quick review of this PR?

File: src/indieweb_utils/feeds/discovery.py
Comment (jamesvandyne) [inline src/indieweb_utils/feeds/discovery.py]: Use a proper type, instead of a class.

```suggestion
def discover_h_feed(url: str) -> Dict:
```

File: src/indieweb_utils/feeds/discovery.py
Comment (jamesvandyne) [reply src/indieweb_utils/feeds/discovery.py]: @capjamesg We should apply this suggestion (or something similar) before merge to prevent a potential crasher.

File: src/indieweb_utils/feeds/discovery.py
Comment (jamesvandyne) [inline src/indieweb_utils/feeds/discovery.py]: This also needs to have a similar conditional like the list compression above it to prevent a potential crash.

**Review by jamesvandyne — APPROVED** [review]
LGTM w/ suggestions

**Comment by capjamesg** [comment]
Thank you for your suggestions @jamesvandyne. I have just incorporated them into my latest commit. Pending approval on all CI actions, I will merge this into main.

**Pull request merged**
